File: native/ios/Sources/Sfiora/NfcOperationCoordinator.swift
import Foundation

enum NfcOperationKind {
    case read
    case write
}

struct NfcOperationLease: Equatable {
    fileprivate let identifier: UUID
    let kind: NfcOperationKind
}

/// Serializes NFC read and write sessions in the current process.
enum NfcOperationCoordinator {
    private static let lock = NSLock()
    private static var activeLease: NfcOperationLease?

    static func acquire(_ kind: NfcOperationKind) -> NfcOperationLease? {
        lock.lock()
        defer { lock.unlock() }
        guard activeLease == nil else {
            return nil
        }
        let lease = NfcOperationLease(identifier: UUID(), kind: kind)
        activeLease = lease
        return lease
    }

    static func release(_ lease: NfcOperationLease?) {
        guard let lease else {
            return
        }
        lock.lock()
        defer { lock.unlock() }
        guard activeLease?.identifier == lease.identifier else {
            return
        }
        activeLease = nil
    }
}

File: native/android/sfiora/src/main/java/com/sandrox/sfiora/NfcOperationCoordinator.java
package com.sandrox.sfiora;

import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicReference;

/** Serializes NFC read and write sessions in the current process. */
final class NfcOperationCoordinator {
    enum Kind {
        READ,
        WRITE
    }

    static final class Lease {
        private final UUID identifier;
        private final Kind kind;

        private Lease(Kind kind) {
            identifier = UUID.randomUUID();
            this.kind = kind;
        }

        Kind getKind() {
            return kind;
        }
    }

    private static final AtomicReference<Lease> ACTIVE =
            new AtomicReference<>();

    private NfcOperationCoordinator() {
    }

    static Lease tryAcquire(Kind kind) {
        Lease lease = new Lease(Objects.requireNonNull(kind, "kind is required"));
        return ACTIVE.compareAndSet(null, lease) ? lease : null;
    }

    static void release(Lease lease) {
        if (lease != null) {
            ACTIVE.compareAndSet(lease, null);
        }
    }
}
